test/helpers/position_assertions.cc

```diff
@@ -851,4 +854,81 @@ string HoverAssertion::toString() const {
     return fmt::format("hover: {}", message);
 }
 
+shared_ptr<ApplyCodeActionAssertion> ApplyCodeActionAssertion::make(string_view filename, unique_ptr<Range> &range,
+                                                                    int assertionLine, string_view assertionContents,
+                                                                    string_view assertionType) {
+    static const regex titleVersionRegex(R"(^\[(\w+)\]\s+(.*?)$)");
+
+    smatch matches;
+    string assertionContentsString = string(assertionContents);
+    if (regex_search(assertionContentsString, matches, titleVersionRegex)) {
+        string version = matches[1].str();
+        string title = matches[2].str();
+        return make_shared<ApplyCodeActionAssertion>(filename, range, assertionLine, title, version);
+    }
+
+    ADD_FAILURE_AT(string(filename).c_str(), assertionLine + 1)
+        << fmt::format("Found improperly formatted apply-code-action assertion. Expected apply-code-action "
+                       "[version] code-action-title.");
+    return nullptr;
+}
+ApplyCodeActionAssertion::ApplyCodeActionAssertion(string_view filename, unique_ptr<Range> &range, int assertionLine,
+                                                   string_view title, string_view version)
+    : RangeAssertion(filename, range, assertionLine), title(string(title)), version(string(version)) {}
+
+string ApplyCodeActionAssertion::toString() const {
+    return fmt::format("apply-code-action: [{}] {}", version, title);
+}
+
+void ApplyCodeActionAssertion::check(const UnorderedMap<std::string, std::shared_ptr<core::File>> &sourceFileContents,
+                                     unique_ptr<CodeAction> &codeAction, string_view testName, string_view fileUri) {
+    auto expectedUpdatedFilePath = fmt::format("{}.{}.rbedited", testName, version);
+    string expectedEditedFileContents;
+    try {
+        expectedEditedFileContents = FileOps::read(expectedUpdatedFilePath);
+    } catch (FileNotFoundException e) {
+        ADD_FAILURE() << fmt::format("Missing {} which should contain test file after applying code actions.",
+                                     expectedUpdatedFilePath);
+        return;
+    }
+
+    auto it = sourceFileContents.find(filename);
+    if (it == sourceFileContents.end()) {
+        ADD_FAILURE() << fmt::format("Unable to find referenced source file `{}`", filename);
+        return;
+    }
+
+    auto &file = it->second;
+    for (auto &c : *codeAction->edit.value()->documentChanges) {
+        string actualEditedFileContents = string(file->source());
+        EXPECT_EQ(c->textDocument->uri, fileUri)
+            << fmt::format("Requested code action for {}, but received edits in {}", fileUri, c->textDocument->uri);
+
+        // First, sort the edits by increasing starting location and verify that none overlap.
+        fast_sort(c->edits, [](const auto &l, const auto &r) -> bool { return cmpRanges(*l->range, *r->range); });
```

Comment:
`fast_sort` expects a boolean return value in the lambda, but `cmpRanges` returns an int. So this line isn't going to work properly right now.

Reply:
(This bug has bitten me many times, and is painful to diagnose.)

Reply:
Completely missed this. Thanks for the catch.